Test webrtc/content-security-policy integration

zenhack opened the pull request:
...as specified in:

- https://github.com/w3c/webappsec-csp/pull/457
- https://github.com/w3c/webrtc-extensions/pull/81

---

This is a work in progress; in particular:

- Right now all of the "allow" tests fail with:

```
  ▶ ERROR [expected OK] /content-security-policy/webrtc/webrtc-allowed-nopolicy.html
  └   → Unhandled rejection: Unknown ufrag (3a846ca3)
```

If I silence that one rejection they pass (and the "blocked" tests still fail), but I don't know the right approach to make sure that event doesn't fire before we can handle it.

- I'm very fuzzy on whether this is actually the right approach; right now this tries to monitor the ice agent's transitions and make sure it goes straight from new to failed, but my grasp of the intricacies of webrtc is still somewhat tenuous, so I would not be surprised if the logic for this was wrong in some critical way.

I would appreciate feedback.

cc: @jan-ivar @alvestrand @annevk @antosart

zenhack committed:
Test webrtc/content-security-policy integration

...as specified in:

- https://github.com/w3c/webappsec-csp/pull/457
- https://github.com/w3c/webrtc-extensions/pull/81

zenhack commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,62 @@
+    // or "failed."
+    const pcFailed = (pc) => {
+        pc.onicecandidate = ({candidate}) => {
+            pc.addIceCandidate(candidate);
```
To be clear, the unhandled rejection is happening here.

jan-ivar replied:
This is calling addIceCandidate on the wrong pc. pc1 should call it on pc2 and vice versa (this being a shortcut-ed signaling dance). See suggestion below.

antosart commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,62 @@
+// blocked on both sides and "inconsistent" in the even that the
+// result is the same on both sides (should never happen).
+async function tryConnect() {
+    const iceServers = [{urls: "stun:stun.l.google.com:19302"}];
```
I'm not sure what the code below does, but I think WPTs should not depend on external services like google.com

zenhack replied:
It seems like in principle we should be able to avoid this, since the stun server is typically for NAT traversal. If I just set this to an empty list though I get:

```
  │   → Unhandled rejection: No remoteDescription.
  │ 
  └ pcFailed/pc.onicecandidate@http://web-platform.test:8000/content-security-policy/webrtc/webrtc.js:17:1
```

...on two of the tests (instead of the ufrag error). Maybe @alvestrand or @jan-ivar can shed some on the right way to do this; I feel like I'm kindof fumbling around piecing this together.

zenhack replied:
Removed.

antosart reviewed (commented):
Sorry, I don't have enough knowledge of webrtc to review the code. Hope some of the other reviewers does.

zenhack commented:
@jan-ivar, @alvestrand, any thoughts?

zenhack commented:
Ping?

jan-ivar commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,62 @@
+    // Creating a data channel is necessary to induce negotiation:
+    const channel = pc1.createDataChannel('test');
+
+    // Usual webrtc signaling dance:
```
```suggestion
    // Usual webrtc signaling dance:
    pc1.onicecandidate = ({candidate}) => pc2.addIceCandidate(candidate);
    pc2.onicecandidate = ({candidate}) => pc1.addIceCandidate(candidate);
```

jan-ivar commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,62 @@
+
+// Creates two RTCPeerConnection and tries to connect them. Returns
+// "allowed" if the connection is permitted, "blocked" if it is
+// blocked on both sides and "inconsistent" in the even that the
+// result is the same on both sides (should never happen).
```
```suggestion
// blocked on both sides and "inconsistent" in the event that the
// result is not the same on both sides (should never happen).
```

jan-ivar commented:
@zenhack That should also take care of the "Unknown ufrag" errors. Otherwise the tests look good, thanks for writing them!

jan-ivar commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,62 @@
+            pc.addIceCandidate(candidate);
+        };
+        return new Promise((resolve, _reject) => {
+            pc.onicegatheringstatechange = (e) => {
+                if(pc.iceGatheringState === "complete") {
+                    resolve(pc.iceConnectionState === "failed");
+                } else if(pc.iceConnectionState !== "new") {
+                    resolve(false);
+                }
+            };
```
One more issue:

While it's correct to look at iceConnectionState, I think we need to listen to `pc.oniceconnectionstatechange` here, not `pc.onicegatheringstatechange`, oherwise iceConnectionState is not guaranteed to have been updated yet.

The timing between updates to the two states isn't super-clear in the spec (but I believe it is specified): Most of the time, the two states are timing independent, but for [failed](https://w3c.github.io/webrtc-pc/#dom-rtcicetransportstate-failed), iceConnectionState is causally dependent on multiple things, one of them being _"finished gathering"_, gatheringState is updated ([often way](https://github.com/w3c/webrtc-pc/issues/2698#issuecomment-987443817)) ahead of iceConnectionState.

If we want to test that the PAC timer is bypassed on CSP block (which it should be), we'd need to test that differently somehow, e.g. timeout after 1 second.

zenhack replied:
Done.

zenhack committed:
Fix typos in comment.

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

zenhack committed:
Fix ice candidate exchange in CSP webrtc tests.

we should be passing each candidate to the *other* pc.

Tests now behave as expected.

zenhack committed:
CSP webrtc tests: listen for connection state change, not gathering.

See https://github.com/web-platform-tests/wpt/pull/32914#discussion_r831320820

zenhack committed:
CSP webrtc tests: drop unnecessary stun server.

zenhack commented:
Ok, I think I have addressed everything.

jan-ivar commented on content-security-policy/webrtc/webrtc.js:
```diff
@@ -0,0 +1,60 @@
+    const pcFailed = (pc) => {
+        return new Promise((resolve, _reject) => {
+            pc.oniceconnectionstatechange = (e) => {
+                if(pc.iceGatheringState === "complete") {
+                    resolve(pc.iceConnectionState === "failed");
+                } else if(pc.iceConnectionState !== "new") {
+                    resolve(false);
+                }
```
The following should suffice, and align better with the code comment above (since `"new"` is the initial state):
```suggestion
                resolve(pc.iceConnectionState == "failed");
```

zenhack committed:
webrtc csp: simplify state checking

"new" is the initial state.

Co-authored-by: Jan-Ivar Bruaroey <jan-ivar@users.noreply.github.com>

zenhack commented:
@annevk, @alvestrand, ping?

youennf commented:
LGTM as well.
It might be cool to add a test with two iframes, one being blocked and the other one not being blocked.
In any case, can we merge it as is?

jan-ivar commented:
@annevk and @antosart the WebRTC parts here look good to us. Can it be merged?

> It might be cool to add a test with two iframes, one being blocked and the other one not being blocked.

That might take some more work, like postMessage between the iframe. A follow-up issue might be good for that.

jan-ivar commented:
Based on [and earlier comment](https://github.com/web-platform-tests/wpt/pull/32914#pullrequestreview-889386074) from @antosart and @annevk's [approval](https://github.com/w3c/webappsec-csp/pull/457#pullrequestreview-881650648) of the PR this is testing, I think we're good to merge.
